File: leetcode/algorithms/string/reverse-vowels-of-a-string.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Vowels of a String"
*/

/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function(s) {
    const vowelSet = new Set(['a','e','i','o','u','A','E','I','O','U']);
    const vowels = [],consonants = [],vowelIndices = [];
    for(let i=0;i<s.length;i++){
        if(vowelSet.has(s[i])){
            vowels.push(s[i]);
            vowelIndices.push(i);
        } else {
            consonants.push(s[i]);
        }
    }
    vowels.reverse();
    let newString="",v=0,c=0;
    for(let i=0;i<s.length;i++){
        if(vowelIndices[v]===i){
            newString+=vowels[v];
            ++v;
        } else {
            newString+=consonants[c];
            ++c;
        }
    }
    return newString;
};

File: leetcode/algorithms/string/reverse-vowels-of-a-string.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Vowels of a String"
*/

class Solution {
    
    private HashSet<Character> VOWELS = new HashSet<Character>(){{
        add('a');
        add('A');
        add('e');
        add('E');
        add('i');
        add('I');
        add('o');
        add('O');
        add('u');
        add('U');
    }};
    
    private boolean isVowel(char c){
        return VOWELS.contains(c);
    }
    
    
    
    public String reverseVowels(String s) {
        
        int low=0,high=s.length()-1;
        
        StringBuilder revBuilder = new StringBuilder(s);
        
        while(high>low){
            boolean isLowVowel = isVowel(s.charAt(low)), isHighVowel = isVowel(s.charAt(high)); 
            if(!isLowVowel){
                ++low;
            }
            if(!isHighVowel){
                --high;
            }
            
            if(isLowVowel && isHighVowel){
                char temp = s.charAt(low);
                revBuilder.setCharAt(low,s.charAt(high));
                revBuilder.setCharAt(high,temp);
                ++low;
                --high;
            }
        }
        
        return revBuilder.toString();
        
        
    }
}